packages/ui/src/components/Visualization/ContextToolbar/ExportDocument/ExportDocument.tsx

```diff
@@ -0,0 +1,159 @@
+import {
+  Button,
+  Form,
+  FormGroup,
+  ModalBody,
+  ModalHeader,
+  Split,
+  SplitItem,
+  TextInput,
+  Toolbar,
+  ToolbarItem,
+} from '@patternfly/react-core';
+import { DownloadIcon, FileIcon } from '@patternfly/react-icons';
+import { Modal, ModalVariant } from '@patternfly/react-core/deprecated';
+import { DocumentationService } from '../../../../services/documentation.service';
+import Markdown from 'react-markdown';
+import remarkGfm from 'remark-gfm';
+import { Element } from 'hast';
+import { FunctionComponent, useContext, useEffect, useState } from 'react';
+import './ExportDocument.scss';
+import { EntitiesContext, VisibleFlowsContext } from '../../../../providers';
+import { markdownComponentMapping } from './MarkdownComponentMapping';
+import { EntitiesMenu } from './EntitiesMenu';
+import { DocumentationEntity } from '../../../../models/documentation';
+
+export const ExportDocument: FunctionComponent = () => {
+  const fileNameBase = 'route-export';
+  const { entities, visualEntities } = useContext(EntitiesContext)!;
+  const { visibleFlows, visualFlowsApi } = useContext(VisibleFlowsContext)!;
+
+  const [isModalOpen, setIsModalOpen] = useState<boolean>(false);
+  const [markdownText, setMarkdownText] = useState<string>('');
+  const [flowImageBlob, setFlowImageBlob] = useState<Blob>();
+  const [downloadFileName, setDownloadFileName] = useState<string>(fileNameBase + '.zip');
+
+  const [documentationEntities, setDocumentationEntities] = useState<DocumentationEntity[]>([]);
+
+  const onUpdateDocumentationEntities = (documentationEntities: DocumentationEntity[]) => {
+    documentationEntities.forEach((docEntity) => {
+      if (
+        docEntity.isVisualEntity &&
+        ((docEntity.isVisible && !visibleFlows[docEntity.entity!.id]) ||
+          (!docEntity.isVisible && visibleFlows[docEntity.entity!.id]))
+      ) {
+        visualFlowsApi.toggleFlowVisible(docEntity.entity!.id);
+      }
+    });
+
+    setDocumentationEntities([...documentationEntities]);
+  };
+
+  useEffect(() => {
+    setTimeout(async () => {
+      const imageBlob = await DocumentationService.generateFlowImage();
+      if (!imageBlob) {
+        console.error('Failed to generate flow diagram image');
+        return;
+      }
+      setFlowImageBlob(imageBlob);
+      const imageUrl = window.URL.createObjectURL(imageBlob);
+      if (!imageUrl) {
+        console.error('Failed to create image URL');
+        return;
+      }
+
+      const md = DocumentationService.generateMarkdown(documentationEntities, imageUrl);
+      setMarkdownText(md);
+    });
+  }, [documentationEntities, visibleFlows]);
```

Comment:
:eyes: what happened? :D `useEffect` it's not very friendly with `async` functions, as it just executes them without waiting for the completion

Reply:
Can you ellaborate? if it's about the `setTimeout`, it's a workaround for this issue - https://github.com/reactwg/react-18/discussions/128
refferred from an official doc - https://react.dev/reference/react/useEffect#caveats

Reply:
The symptom was that the image capturing was happening "before" the Canvas is re-rendering, then the captured image shows the previous status of the flows visibility